libass/ass_parse.c

```diff
@@ -232,6 +232,7 @@ static int parse_vector_clip(ASS_Renderer *render_priv,
  */
 char *parse_tags(ASS_Renderer *render_priv, char *p, char *end, double pwr)
 {
+start:
```

Comment:
Not loving this very much. Makes it a bit hard to reason about whether it really makes progress every time it gets here. (Seems `p` is updated as parsing progresses...) Also, goto is ugly.

Maybe this could be solved by another indirection by making a function call parse_tags itself in a loop?

Reply:
I don’t think another function calling this in a loop would work, because the function’s arguments—`pwr` and later `nested`—need to be updated. Although I guess we could change them into pointers (in the inner function’s argument list only) so that the inner function call can update them.

One option to avoid the `goto` would be to use an actual tail call and rely on the compiler to optimize it, which I see you’ve mentioned on IRC. I think it should be safe but may look a bit weird when we go back and reread this code years later.

I thought of phrasing the loop like this:
```c
while (p < end) {
    ...
    p = q;
}
```
possibly with a `goto` _within_ the loop, but then I don’t see how to manage the `continue`s in tags such as `\fade`, which should execute the `p = q` assignment while the pseudo-tail-call in `\t` shouldn’t execute it. Either all those `continue`s need to be replaced with `goto`s—which feels wrong because the fewer `goto`s, the better—(and the `goto` in `\t` replaced with an actual `continue`), or the same effect can be achieved by wrapping the loop body in a `switch (0) { case 0: }` and using `break` instead of `goto`—which feels even more hackish—or a Boolean flag can be used to determine whether the `p = q` assignment should run, which I don’t like either.

Alternatively, perhaps I can just reassign `q = p` in the pseudo-tail-call path and let the loop-increment statement `p = q` happen as usual:
```c
for (char *q; p < end; p = q) {
    ...
        if (args[cnt].end < end) {
            ...
        } else {
            pwr = k;
            nested = true;
            q = p;
            continue;  // or not, because there’s no code after this anyway
        }
    ...
}
```

Reply:
The function could pass in those arguments as pointers (or as a pointer to a struct).

Relying on tail call optimization if the goal is to avoid recursion is probably not a good idea, regrettably.

I think the goal should be to keep this somewhat readable - your last suggestion doesn't look too bad?